### DISCUSSION TITLE
Clientside E2

### DISCUSSION THREAD

**Discussion by vurvdev** [wiremod/wire#2909] [Suggestions]
There's been a lot of talk about it in the discord, but its about time it got a formal discussion

The proposal to add clientside E2 comes from the painful limitations of networking that has plagued stuff like file functions, http functions and especially EGP.

I don't imagine this will happen anytime soon since this needs to be discussed heavily (and I want E2 to be more heavily optimized before this happens).

## Syntax

This can go a lot of ways, most popular seems to be the EA3/Expressive syntax

```ts
client {
    // do stuff
}

server {
    // do stuff
}
```

I'll also put this on the table as an option to avoid breaking typescript syntax (I'd add an editor autocomplete option to do it for you). Don't think people like this.

```ts
{
    "use server"
    // do stuff
}

{
    "use client"
     // do stuff
}
```

Of course `@server` and `@client` directives will exist. `@server` will be default.

## Changes

This would largely make EGP, as it currently stands, obsolete.

With clientside E2 there's a lot of possibilities for clientside rendering.

Of course there will be raw rendering events and functions, but I'd like to imagine an EGPv4 being implemented with JSX syntax or something (similar to wire extras' hud)

## Prerequisites

- [ ] Better networking (netstream needs to be pulled into wire likely)
- [ ] Optimizer pass
- [ ] ?

**Comment by Denneisk** [comment]
I think aiming to follow TS to a T will only provide more trouble than it's worth. My view of E2 has been to be a domain-specific extension of TS (with a lot of legacy baggage).

I think defining an API for defining CS/SS code is important. It could just be something like prefixing files with `cl_` and `sv_` for all I care. I think the extension system needs a rework before that can happen. See also #2759 and #3071
